### ISSUE TITLE
FinalLocalVariable: false positive when variable is assigned multiple times

### ISSUE THREAD

**Issue by rnveach** [checkstyle/checkstyle#3006]
````
$ cat TestClass.java
public class Test {
    void method() {
        final boolean some_condition = true;
        int i;

        if (some_condition) {
            i = 1;

            // work i
        }

        i = 2;
        // work i in another way
    }

    void method2() {
        final boolean some_condition = true;
        int i;

        {
            i = 2;
        }

        // work i

        if (some_condition) {
            i = 1;

            // work i in another way
        }
    }
}

$ cat TestConfig.xml
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
          "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
          "http://www.puppycrawl.com/dtds/configuration_1_3.dtd">

<module name="Checker">
    <property name="charset" value="UTF-8"/>

    <module name="TreeWalker">
<module name="FinalLocalVariable" />
    </module>
</module>

$ java -jar checkstyle-6.16-all.jar -c TestConfig.xml TestClass.java
Starting audit...
[ERROR] TestClass.java:4:13: Variable 'i' should be declared final. [FinalLocalVariable]
[ERROR] TestClass.java:18:13: Variable 'i' should be declared final. [FinalLocalVariable]
Audit done.
Checkstyle ends with 2 errors.
````

The examples are a trimmed down version of the cases from my code.
I was expecting no errors. `i` cannot be declared `final` in any of the cases since it could be assigned a value multiple times depending on the result of the condition. Eclipse gives me an error when I try to make the variables `final`.

**Comment by romani** [comment]
@Bhavik3 , can you help us one more time with this Check?

**Comment by kaj** [comment]
The i inside the condition is logically not the same variable as the i outside the condition (as the variable is assigned before any use outside the condition, the value from inside the condition does not leak outside), so the code for `method` should be changed like this:

    void method() {
        final boolean some_condition = true;
        if (some_condition) {
            final int j = 1;
            // work j
        }

        final int i = 2;
        // work i in another way
    }

Roughly the same applies true for `method2`.

**Comment by MEZk** [comment]
@kaj 
I think his intension was to point that he does not want to use different names for variables or add a new variable.

**Comment by kaj** [comment]
@MEZk , I think the point of the test is to point out that xe should.  Whether to use that test or not in a project is configurable.  :-)

**Comment by Bhavik3** [comment]
@MEZk 
Here point is not to change the code. But for the same condition solve the False Positive.

As far as I can think, this type of problem arrives when you use only if condition and not if_else.
So, need to change the algorithm little bit in a way it can handle also only If condition.

**Comment by romani** [comment]
this Check is not not in a position to demand refactoring from user, it demand only addition of "final" modifier to existing declaration.

Issue is 100% valid. UTs input files should have exactly this test case (any extra tests are good too).

**Comment by Bhavik3** [comment]
@MEZk 

I think you need to change shouldUpdateUninitializedVariables(DetailAST ast) method in a way that it returns true (when leaving SLIST token of IF block) for the following case
```
        int i;
        if (some_condition) {
            i = 1;
        }
```
 And return false (when leaving SLIST token of IF block) for the following case
```
        int i;
        if (some_condition) {
            i = 1;
        } else {
            i = 2;
        }
```

**Comment by MEZk** [comment]
Few more False Positives:
1)
```java
    void foo2() {
        final boolean some_condition = true;
        int i;
        if (some_condition) {
            i = 1;
        }
        else {

        }
        i = 2;
    }
```

2)
```java

    void foo6() {
        final boolean some_condition = true;
        int i;

        {
            i = 2;
        }

        if (some_condition) {
            i = 1;
        }
        else {
            i = 6;
        }
    }
```

**Comment by archmageirvine** [comment]
I'm also seeing a bunch of false-positives from FinalLocalVariable in conjunction with switch statements.  For example "b" in the following function:

` private static void combine(final int[][] m, final int x, final int y, final int d, final int[] regionSizes) {
    // Get label for chosen cell and target cell
    final int a = m[y][x] & NUM_MASK;
    int b;
    switch (d) {
    case TOP:
      b = m[y - 1][x];
      break;
    case LEFT:
      b = m[y][x - 1];
      break;
    case RIGHT:
      b = m[y][x + 1];
      break;
    default: // BOTTOM
      b = m[y + 1][x];
      break;
    }
    b &= NUM_MASK;
    assert a != b;
    regionSizes[b >>> NUM_SHIFT] += relabelRegion(m, a, b);
    removeRedundantWalls(m, b);
  }
`

**Comment by romani** [comment]
@MEZk, please recheck your fix on this code

**Comment by romani** [comment]
Fix is merged

**Comment by MEZk** [comment]
@romani 
Example which is given by @archmageirvine causes false positive before and after my fix. Should I open separate issue or try to fix the problem in scope of the current issue?

**Comment by romani** [comment]
Please open separate issue
